lib/src/source/hosted.dart

```diff
@@ -152,55 +154,117 @@ class HostedSource extends Source {
     return {'name': name, 'url': url.toString()};
   }
 
+  @override
+  dynamic serializeDescription(String containingPath, description) {
+    final desc = _asDescription(description);
+    return _descriptionFor(desc.packageName, desc.uri);
+  }
+
   @override
   String formatDescription(description) =>
-      'on ${_parseDescription(description).last}';
+      'on ${_asDescription(description).uri}';
 
   @override
   bool descriptionsEqual(description1, description2) =>
-      _parseDescription(description1) == _parseDescription(description2);
+      _asDescription(description1) == _asDescription(description2);
 
   @override
-  int hashDescription(description) => _parseDescription(description).hashCode;
+  int hashDescription(description) => _asDescription(description).hashCode;
 
   /// Ensures that [description] is a valid hosted package description.
   ///
-  /// There are two valid formats. A plain string refers to a package with the
-  /// given name from the default host, while a map with keys "name" and "url"
-  /// refers to a package with the given name from the host at the given URL.
+  /// Simple hosted dependencies only consist of a plain string, which is
+  /// resolved against the default host. In this case, [description] will be
+  /// null.
+  ///
+  /// Hosted dependencies may also specify a custom host from which the package
+  /// is fetched. There are two syntactic forms of those dependencies:
+  ///
+  ///  1. With an url and an optional name in a map: `hosted: {url: <url>}`
+  ///  2. With a direct url: `hosted: <url>`
   @override
-  PackageRef parseRef(String name, description, {String containingPath}) {
-    _parseDescription(description);
-    return PackageRef(name, this, description);
+  PackageRef parseRef(String name, description,
+      {String containingPath, LanguageVersion languageVersion}) {
+    return PackageRef(
+        name, this, _parseDescription(name, description, languageVersion));
   }
 
   @override
   PackageId parseId(String name, Version version, description,
       {String containingPath}) {
-    _parseDescription(description);
-    return PackageId(name, this, version, description);
+    // Old pub versions only wrote `description: <pkg>` into the lock file.
+    if (description is String) {
+      assert(description == name);
+      return PackageId(
+          name, this, version, HostedDescription(name, defaultUrl));
+    }
+
+    final serializedDescription = (description as Map).cast<String, String>();
+
+    return PackageId(
+      name,
+      this,
+      version,
+      HostedDescription(serializedDescription['name'],
+          Uri.parse(serializedDescription['url'])),
+    );
   }
 
+  HostedDescription _asDescription(desc) => desc;
```

Comment:
What do we need this for? I'm guessing it's a left-over that we can now remove :D

Reply:
The description we get from package names or refs is always `dynamic`, so this is essentially a cast. I'll make the code more obvious by spelling out `desc as HostedDescription`. We could also inline the cast into all usages.